src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs

```diff
@@ -610,44 +610,54 @@ public async ValueTask DisposeAsync()
-                    clusterTransactionCommand.FromBackup = true;
+                var clusterTransactionCommand = new ClusterTransactionCommand(_database.Name, _database.IdentityPartsSeparator, topology, parsedCommands, options, raftRequestId);
+                clusterTransactionCommand.FromBackup = true;
 
-                    var clusterTransactionResult = await _database.ServerStore.SendToLeaderAsync(clusterTransactionCommand);
-                    for (int i = 0; i < _clusterTransactionCommands.Length; i++)
-                    { 
-                        _clusterTransactionCommands[i].Document.Dispose();
-                        _clusterTransactionCommands[i].OriginalChangeVector.Dispose();
-                    }
-             
-                    _clusterTransactionCommands.Clear();
-                    _documentContextHolder.Reset();
+                var clusterTransactionResult = await _database.ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+                for (int i = 0; i < _clusterTransactionCommands.Length; i++)
+                {
+                    _clusterTransactionCommands[i].Document.Dispose();
+                    _clusterTransactionCommands[i].OriginalChangeVector.Dispose();
+                }
+
+                _clusterTransactionCommands.Clear();
+                _documentContextHolder.Reset();
 
-                    if (clusterTransactionResult.Result is List<string> errors)
-                        throw new ConcurrencyException($"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors)}"); //TODO
+                if (clusterTransactionResult.Result is List<string> errors)
+                    throw new ConcurrencyException($"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors)}");
 
-                    await _database.ServerStore.Cluster.WaitForIndexNotification(clusterTransactionResult.Index);
-                    await _database.ClusterTransactionWaiter.WaitForResults(taskId, _token);
+                await _database.ServerStore.Cluster.WaitForIndexNotification(clusterTransactionResult.Index);
+
+                //When restoring from snapshot the database doesn't exist yet and we cannot rely on the DocumentDatabase to execute the database cluster transaction commands
+                using (_database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext writeContext))
+                using (writeContext.OpenWriteTransaction())
+                using (var rawDatabaseRecord = _database.ServerStore.Cluster.ReadRawDatabaseRecord(writeContext, _database.Name, out _))
+                {
+                    if (rawDatabaseRecord.DatabaseState == DatabaseStateStatus.RestoreInProgress)
+                    {
+                        await _database.ExecuteClusterTransaction(writeContext);
```

Comment:
What is the scenario where this happens? You are reading _existing commands_, no?
But if we are during restore, how could they _get_ there? 

Also, what about the cluster wide effects?

Reply:
In the usual import process, we create those on import, IIRC, no? Are you handling that scenario here?

Reply:
> What is the scenario where this happens? You are reading _existing commands_, no? But if we are during restore, how could they _get_ there?

In `DatabaseDestination` for each `compare exchange`, we check if it is an atomic guard and later (end of a batch or dispose) we send a cluster transaction command with the documents to create the `compare exchange`s.

> Also, what about the cluster wide effects?

What do you mean? The `ClusterStateMachine` adds the cluster transaction database commands to the `ClusterStateMachine.TransactionCommands` table and only the local database reads it and responsible to execute them.
(despite of cleaning handled by the cluster state machine `ClusterStateCleanUp`)
Do I miss something?

> In the usual import process, we create those on import, IIRC, no? Are you handling that scenario here?

"those" stand for the `compare exchange`s? We create them by sending cluster transaction command with `put` for the relevant documents.

https://github.com/haludi/ravendb/blob/RavenDB-17174/src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs#L572

Reply:
So I may be missing something. Why do we need to do anything here at all? Shouldn't the transaction writes to the docs happen when we finish loading the database? Or is it that it is already started, and we missed that section in the startup?

Reply:
We can just not wait for the cluster transaction commands to be completed.
But then you can get confirmed for import to be completed while the cluster transaction database commands are not.
The document data will stay the same since we just putting it with the same data to create the compare exchange atomic guard but the change vector and the modified time will change.
If we don't care for it at all or just don't care in restore from a snapshot I can simplify the code.

Reply:
I still don't quite understand the difference here between normal and snapshot restore here.
Why is that work as expected in the normal restore but not in the snapshot?

Reply:
Because the document database the restore uses is not fully initialized and so the task responsible for executing the cluster transaction database command is not running.
https://github.com/haludi/ravendb/blob/RavenDB-17174/src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs#L283
https://github.com/haludi/ravendb/blob/RavenDB-17174/src/Raven.Server/Documents/DocumentDatabase.cs#L302

Even though I can include the task (responsible for executing the cluster transaction database) in the partial initialization, the state for which commands were done is an in-memory state, and the persistence has been done by the cleaner that cleans the already done commands. 
The cleaner works on database records (which does not exist for the partially initialized database).
If we don't clean the database cluster transaction table the commands will run again in the database load.
https://github.com/haludi/ravendb/blob/RavenDB-17174/src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs#L194
https://github.com/haludi/ravendb/blob/RavenDB-17174/src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs#L375

So I think the simpler way is just to execute and clean the commands directly from the DatabaseDestination.